### PR TITLE
Make sure we stop resolves triggered by a browse when the browse stops on Darwin.

### LINKED ISSUE #24074: Make sure we clean up resolves started by a browse if they have not completed when the browse completes.
Right now Darwin browse relies on the platform dns middle layer to do a resolve on the browse results.  That means we get DNSServiceResolve calls that no one ever cancels.

We should just do the resolves internally using the same service ref as the browse, so canceling the browse cancels the resolves as well.  Need to make sure all the various refcounting is correct in the caller and that we don't drop the service ref too early inside the Darwin code.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
Without this change, if there is a PTR record that matches whatever we are browsing but no corresponding SRV record, we would end up leaking a resolve forever.

Tested by modifying minimal mdns SrvResponder::AddAllResponses to no-op instead of actually adding any responses, then trying to commission the device running the modified minimal mdns.  Without this change, when the browse stops the resolves it triggered keep going.  With this change, termination of the browse also terminates the resolves.

Fixes https://github.com/project-chip/connectedhomeip/issues/24074

### CODE REVIEW

**Comment by vivien-apple** [comment]
The `static sContextDispatchingSuccess` is ... 
Sadly I don’t have anything better to propose since I don’t have a good idea about how to to keep track of the original browse using the platform API abstraction :disappointed: 
And so the magic happens because the context got cleared when `ChipDnssdStopBrowse` happens ?

**Review by vivien-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
>  And so the magic happens because the context got cleared when ChipDnssdStopBrowse happens ?

Exactly.  That shuts down all the things sharing that connection.

**Review by andy31415 — APPROVED** [review]

File: src/platform/Darwin/DnssdImpl.cpp [hunk 1]
```diff
@@ -455,8 +467,21 @@ CHIP_ERROR ChipDnssdStopBrowse(intptr_t browseIdentifier)
...
     // results we have queued up in the BrowseContext, if any.  In practice
     // there shouldn't be anything there long-term anyway.
-    auto browseCtx = static_cast<BrowseContext *>(ctx);
-    browseCtx->Finalize(kDNSServiceErr_Timeout);
+    //
+    // Make sure to time out all the resolves first, before we time out the
+    // browse (because after the latter the serviceRefs on the resolves will not
+    // really be usable).
+    std::vector<GenericContext *> resolves;
+    MdnsContexts::GetInstance().FindAllMatchingPredicate(
```
Comment (bzbarsky-apple) [inline src/platform/Darwin/DnssdImpl.cpp:475]: This part needs re-review from @vivien-apple

**Comment by bzbarsky-apple** [comment]
So the previous approach does not work because each ResolveContext expects to get an update without the "more coming" flag and acts on that, but if they are all sharing a connection only one of them will get such an update.

Since we have to track the ResolveContexts anyway to clean them up, I am going to just having each one use a separate connection.

File: src/platform/Darwin/DnssdContexts.cpp [hunk 2]
```diff
@@ -156,6 +156,9 @@ CHIP_ERROR MdnsContexts::Add(GenericContext * context, DNSServiceRef sdRef)
     auto err = DNSServiceSetDispatchQueue(sdRef, chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue());
     if (kDNSServiceErr_NoError != err)
     {
+        // We can't just use our Delete to deallocate the service ref here,
+        // because our context may not have its serviceRef set yet.
+        DNSServiceRefDeallocate(sdRef);
```
Comment (vivien-apple) [inline src/platform/Darwin/DnssdContexts.cpp:161]: Is this still needed ?

**Review by vivien-apple — APPROVED** [review]

File: src/platform/Darwin/DnssdContexts.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/platform/Darwin/DnssdContexts.cpp:161]: I think it's needed in general, yes.  When we get here, we have already allocated the service ref, and if we don't clean it up here nothing will ever clean it up....  So it's not strictly related to the other changes here, just something I found while working on this.

Comment (vivien-apple) [reply src/platform/Darwin/DnssdContexts.cpp:161]: Doing a quick look reveal that you are correct. I wonder if we should clear the context too at https://github.com/project-chip/connectedhomeip/pull/24733/files/e41aedc6fa18fa4fcba4c67734e2a5429c05e61e#diff-d3d1d99d691d0aef90e9de721cad391450d0e28aa0b727d812022acc3755f248R148

Comment (bzbarsky-apple) [reply src/platform/Darwin/DnssdContexts.cpp:161]: So up above where we check for `context == nullptr`?  Yes, we should, if sdRef is not null...  Will do that.

**Pull request merged**